### DISCUSSION TITLE
Preço e prazo de fretes pelos correios

### DISCUSSION THREAD

**Discussion by tupizz** [BrasilAPI/BrasilAPI#232] [Ideas]
Seguindo a ideia e problemática apresentada no readme: 

> Acesso programático de informações é algo fundamental na comunicação entre sistemas mas, para nossa surpresa, uma informação tão útil e pública quanto um CEP não consegue ser acessada diretamente por um navegador por conta da API dos Correios não possuir CORS habilitado.

Uma das funções que pessoalmente precisei implementar no servidor, uma vez que conforme dito não conseguimos acessar direto do navegador, é a função retorno do valor de frete, dado o cep de origiem, o cep de destino, e mais algumas informações do produto.

Links:

[Preços e prazos Correios
](https://www.correios.com.br/enviar-e-receber/ferramentas/calculador-remoto-de-precos-e-prazos)

**Comment by brunomarks7** [comment]
Isso facilitaria muito a vida do ecommerce brasileiro. A disponibilidade da API de fretes está longe de 99.9%, principalmente em datas promocionais nacionais, como black friday.

Estou disposto a participar como dev dessa feature.

**Comment by wferreiracosta** [comment]
Correios disponibiliza uma API para o calculo de entrega, calculo de preço da entrega e etc. Poderíamos utilizar.

http://ws.correios.com.br/calculador/CalcPrecoPrazo.asmx

**Comment by RodriAndreotti** [comment]
Cara, esta API tem dado problema direto, um amigo meu utilizou ela em uma loja virtual que ele desenvolveu, ele teve que implementar uma alternativa usando Frenet, pois a API dos correios direto estava dando timeout, ou erro 500 ou algum outro erro qualquer.

E acho impressionante a decadência do serviço, pois algum tempo atrás fiz essa integração e funcionava muito bem, com um uptime muito bom e de uns tempos para cá começou a ficar instável a um ponto que não estava mais confiável a utilização em produção.

**Comment by wferreiracosta** [comment]
Não sabia, obrigado pela informação. Tava achando esquisito que poucas pessoas falavam dela, agora ta explicado kkkk.

**Comment by RodriAndreotti** [comment]
Infelizmente ela tem estado bem instável ultimamente, se houvesse alguma api aberta para isso, como existe a do CEP, seria mais fácil, porém ainda é possível tentar fazer via web scrapping, mas ai qualquer alteração no layout da página dos correios poderia quebrar o módulo.

**Comment by brunomarks7** [comment]
Pessoal, vou colocar comentários que podem ser úteis na discussão:

Eu tenho já instalei alguns ecommerces e utilizo um plugin chamado [Correios para Woocommerce](https://wordpress.org/plugins/woocommerce-correios/) criado pelo Cláudio Sanches da Automattic, que mantém o Woocommerce. 

Sobre minha experiência existe muita intermitência no server dos correios e é comum relatos de clientes que não conseguem calcular fretes. 

Em contrapartida, existe um plugin chamado [Frete offline](https://fernandoacosta.net/produto/frete-tabela-offline/) desenvolvido pelo Fernando Acosta e com cobrança anual e é baseado em planilha com faixas de CEP.

Existe dois serviços que entrega cotações de frete dos correios usando seus próprios servidores. Um deles é o Melhor Envio.  Outro é o Frenet.

Então de fato há formas de desenvolver uma api própria, não sei se com webscraping, precisaria ter consistência para não quebrar em mudanças nos parâmetros dos correios.

Abaixo um exemplo de URL que traz um frente passando a cubagem e peso, cep de origem e destino.
http://ws.correios.com.br/calculador/CalcPrecoPrazo.aspx?nCdServico=04596&sCepDestino=29157416&sCepOrigem=11050250&nVlAltura=6&nVlLargura=11&nVlDiametro=0&nVlComprimento=16&nVlPeso=0,04&nCdFormato=1&sCdMaoPropria=N&nVlValorDeclarado=0&sCdAvisoRecebimento=N&StrRetorno=xml#osds

Há dois parâmetros que podem ser passados para mostrar valores diferentes com base em clientes (donos de loja) que possuam contrato com os Correios.

São eles:

- nCdEmpresa _(número do contrato)_

- sDsSenha _(senha do contrato)_

**Comment by RodriAndreotti** [comment]
@brunomarks7 Lembrando também, que o valor do frete também pode sofrer alterações de acordo o o valor declarado do produto.
O serviço Frenet, se não me engano, trabalha com uma réplica do do banco de dados dos correios, que é atualizada de tempos em tempos por eles, e, quando fazemos um request ao API deles, eles tentam verificar o serviço dos correios em primeiro lugar, se não tiverem sucesso mandam o resultado baseado nesta tabela de contingência deles.
Eles também tem a opção de usar um contrato próprio com os correios, além de ser possível configurar o envio usando o contrato deles, e sendo assim, funcionaria no mesmo estilo do mercado envios.

**Comment by brunomarks7** [comment]
@RodriAndreotti perfeito! O ruim desse processo de tentar consultar nos Correios e usar um servidor próprio como segunda opção de fallback é que geralmente o tempo de resposta dos Correios (apesar do server no Brasil) sempre será mais lerdo, ainda mais se estiver congestionado em datas comerciais. Se não me engano a Frenet tem uma tolerância, um timeout para os serviços dos Correios pra passar a requisitar a API própria.

Qualquer ganho de tempo é muito importante para o ecommerce, mesmo 100ms.

Eu fiquei fascinado com a latência desses servidores da Vercel!

O que pra mim é um mistério é qual melhor forma de obter esses dados, se é via Web Scraping mesmo ou com algum outro método.

**Comment by RodriAndreotti** [comment]
Então, @brunomarks7 !
Imagino eu que a forma menos instável, considerando a atual situação do webservice dos correios, seja o web scrapping, porém o serviço fica completamente suscetível a uma mudança de layout no site deles, e ai, lascou-se... kkk

E usar um serviço, tipo frenet, que tem uma limitação na quantidade de requisições mensais, não sei se é a melhor forma também. :/

Esta é uma discussão que seria interessante aprofundar antes de mergulhar no desenvolvimento deste módulo (que é um módulo muito útil, diga-se de passagem)

**Comment by brunomarks7** [comment]
Perfeito Rodrigo!

Concordo plenamente com você. Ter uma API confiável, com baixa latência e
com um bom cache para cotações de fretes seria um passo importante e que
resultaria em um aumento de vendas em muitos ecommerces.

O Frenet eu nem mencionei no sentido de usar o serviço deles, pois o ideal
é ter menos intermediários possíveis. Mas sim no sentido de investigar o
que essas empresas fazem para ter a base atualizada.

Vamos continuar com as ideias, um abraço!

Em qui., 16 de jul. de 2020 às 13:39, Rodrigo Texeira Andreotti <
notifications@github.com> escreveu:

> Então, @brunomarks7 <https://github.com/brunomarks7> !
> Imagino eu que a forma menos instável, considerando a atual situação do
> webservice dos correios, seja o web scrapping, porém o serviço fica
> completamente suscetível a uma mudança de layout no site deles, e ai,
> lascou-se... kkk
>
> E usar um serviço, tipo frenet, que tem uma limitação na quantidade de
> requisições mensais, não sei se é a melhor forma também. :/
>
> Esta é uma discussão que seria interessante aprofundar antes de mergulhar
> no desenvolvimento deste módulo (que é um módulo muito útil, diga-se de
> passagem)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/filipedeschamps/BrasilAPI/issues/75#issuecomment-659531194>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG3TMMDNIXFOLZM7N4MHFVLR34UL7ANCNFSM4NXRZBHQ>
> .
>


-- 
*Bruno Marques*

**Comment by wesleyburlani** [comment]
Opa pessoal, vivi um pouco dessa dor trabalhando com integração de Marketplace. O caminho seguido geralmente é manter uma planilha estática e consultar dela. De tempo em tempo, geralmente madrugada, atualizar essa planilha numa rotina densa, consultando nos correios todas as faixas de cep (tentando até conseguir consultar). Isso era necessário justamente por causa da latencia indesejada no ambiente dos e-commerces/marketplaces.

**Comment by RodriAndreotti** [comment]
> 
> 
> Opa pessoal, vivi um pouco dessa dor trabalhando com integração de Marketplace. O caminho seguido geralmente é manter uma planilha estática e consultar dela. De tempo em tempo, geralmente madrugada, atualizar essa planilha numa rotina densa, consultando nos correios todas as faixas de cep (tentando até conseguir consultar). Isso era necessário justamente por causa da latencia indesejada no ambiente dos e-commerces/marketplaces.

Mas para isso precisaríamos saber também qual é a regra de negócios para o cálculo do valor, pois o valor do frete varia de acordo com peso, tamanho e valor declarado do objeto.

**Comment by wesleyburlani** [comment]
> > Opa pessoal, vivi um pouco dessa dor trabalhando com integração de Marketplace. O caminho seguido geralmente é manter uma planilha estática e consultar dela. De tempo em tempo, geralmente madrugada, atualizar essa planilha numa rotina densa, consultando nos correios todas as faixas de cep (tentando até conseguir consultar). Isso era necessário justamente por causa da latencia indesejada no ambiente dos e-commerces/marketplaces.
> 
> Mas para isso precisaríamos saber também qual é a regra de negócios para o cálculo do valor, pois o valor do frete varia de acordo com peso, tamanho e valor declarado do objeto.

Você está correto. Esxistem mais dados que são levados em consideração no caclulo. O manual dos correios é disponibilizado em https://correios.com.br/enviar-e-receber/ferramentas/calculador-remoto-de-precos-e-prazos/pdf/manual-de-implementacao-do-calculo-remoto-de-precos-e-prazos/view

dentre eles:
* Codigo da empresa para contrato com correios
* Serviço (PAC, SEDEX...) 
* CEP origem
* CEP destino
* Peso
* Formato da encomenda(Pacote, envelope..)
* Dimensões
* Valor do produto

**Comment by brunomarks7** [comment]
Andei pesquisando sobre cálculo de cubagem dos Correios e encontrei o seguinte:

https://esales.com.br/blog/calculo-de-frete-correios/

Só para documentar pois de repente alguém tem alguma ideia com relação a esse dado. 
Como comentamos, talvez não seja a linha a seguir, pois o web scraping da API rodando periodicamente de fato seja a melhor opção, até para evitar quebras por mudanças nas regras de negócio.

Quero em breve começar a fazer alguns testes de rotinas com múltiplas variáveis de pesos e CEPs diferentes, gravando o retorno da API dos correios.

**Comment by pedrosancao** [comment]
Criei um novo PR #260 para continuar o trabalho feito no #76, tem algumas tarefas pendentes, caso alguém tenha interesse em colaborar.

**Comment by LorhanSohaky** [comment]
Só para eu entender, vocês acham que faz sentido colocar essa rota no BrasilAPI? Pelo que entendo todas as demais rotas funcionam com dados cacheados e podem ser reaproveitadas por outros usuários. Pergunto isso porque sinto um cheiro de que essa rota seria muito específica para cada cliente, o que não tiraria um bom proveito dos caches. E quanto é um bom tempo para se ter de cache do preço e prazo do frete?

**Comment by m1guer** [comment]
algum update? essa rota esta disponivel?

**Reply by RodriAndreotti** [reply]
Então, existe um ponto muito importante que precisa ser pensado para este caso:

A API dos correios só é acessível para clientes que tenham contrato com os correios, sendo assim não teríamos como ter acesso sem ter um CNPJ e um contrato com eles, ou que o usuário da Brasil API o tivesse.

No meu ponto de vista, sendo dessa forma, não faz muito sentido termos essa api do nosso lado, pois apenas seria um proxy para bater na api dos correios, aumentando consideravelmente o tempo de resposta, o que numa cotação de frete geralmente não é desejado.

Eu acredito que o caminho mais fácil para quem precisa de cotação de frete nos correios especificamente, seria usar a própria API deles, que hoje já está bem melhor que antes, já é rest em vez de soap e tem um tempo de resposta razoável.
